8.12: Included build not found in build

### Current Behavior

My current build fails with:

```
Could not determine the dependencies of task ':apollo-kotlin:rootPackageJson'.
> Included build 'apollo-kotlin' not found in build 'apollo-kotlin'. 
```

### Expected Behavior

I expect it to work as in 8.11. Or find a better solution if something changed in 8.12

### Context (optional)

Apologies I haven't found a more minimal reproducer. 

See https://github.com/apollographql/apollo-kotlin/pull/6354 for how I ended up making it work.

### Self-contained Reproducer Project

To reproduce:

```
git clone https://github.com/apollographql/apollo-kotlin
cd apollo-kotlin
git checkout c256f76d0e21db11c5af1baf8a5564d7249697b3
sed -i old 's/11/12/g' gradle/wrapper/gradle-wrapper.properties
./gradlew -p tests build
```

Output:

```
> Configure project :apollo-kotlin:apollo-gradle-plugin
Signing plugin detected. Will automatically sign the published artifacts.

[Incubating] Problems report is available at: file:///Users/martinbonnin/git/apollo-kotlin/tests/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':apollo-kotlin:rootPackageJson'.
> Included build 'apollo-kotlin' not found in build 'apollo-kotlin'. 
```

### Gradle version

8.12

### Gradle version that used to work

8.11.1

--- a/platforms/software/dependency-management/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadata.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadata.java
@@ -17,22 +17,34 @@
 package org.gradle.composite.internal;
 
 import org.gradle.api.artifacts.component.ComponentArtifactIdentifier;
-import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
 import org.gradle.api.tasks.TaskDependency;
 import org.gradle.internal.component.local.model.LocalComponentArtifactMetadata;
 import org.gradle.internal.component.model.IvyArtifactName;
 
 import java.io.File;
 
+/**
+ * Represents an artifact that has its owning component ID overridden.
+ * <p>
+ * This is used when an artifact is referenced from another build, and its owning component id is overridden to its foreign counterpart.
+ * This will go away in Gradle 9.0, when we no longer need to override the owning component id with a foreign identifier.
+ * <p>
+ * This should be better named as {@code OverriddenComponentIdArtifactMetadata} or similar, but Kotlin currently references this internal class.
+ *
+ * @see <a href="https://github.com/JetBrains/kotlin/blame/83a7ecda3eccd378a3882b81d8e83ecc12b3b786/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/targets/js/npm/resolver/KotlinCompilationNpmResolver.kt#L245">Link</a>
+ */
 public class CompositeProjectComponentArtifactMetadata implements LocalComponentArtifactMetadata, ComponentArtifactIdentifier {
-    private final ProjectComponentIdentifier componentIdentifier;
-    private final LocalComponentArtifactMetadata delegate;
-    private final File file;
 
-    public CompositeProjectComponentArtifactMetadata(ProjectComponentIdentifier componentIdentifier, LocalComponentArtifactMetadata delegate, File file) {
-        this.componentIdentifier = componentIdentifier;
+    private final ComponentIdentifier overrideComponentId;
+    private final LocalComponentArtifactMetadata delegate;
+
+    public CompositeProjectComponentArtifactMetadata(
+        ComponentIdentifier overrideComponentId,
+        LocalComponentArtifactMetadata delegate
+    ) {
+        this.overrideComponentId = overrideComponentId;
         this.delegate = delegate;
-        this.file = file;
     }
 
     @Override
@@ -45,8 +57,8 @@ public class CompositeProjectComponentArtifactMetadata implements LocalComponent
     }
 
     @Override
-    public ProjectComponentIdentifier getComponentId() {
-        return componentIdentifier;
+    public ComponentIdentifier getComponentId() {
+        return overrideComponentId;
     }
 
     @Override
@@ -60,8 +72,8 @@ public class CompositeProjectComponentArtifactMetadata implements LocalComponent
     }
 
     @Override
-    public ProjectComponentIdentifier getComponentIdentifier() {
-        return componentIdentifier;
+    public ComponentIdentifier getComponentIdentifier() {
+        return overrideComponentId;
     }
 
     @Override
@@ -76,7 +88,7 @@ public class CompositeProjectComponentArtifactMetadata implements LocalComponent
 
     @Override
     public File getFile() {
-        return file;
+        return delegate.getFile();
     }
 
     @Override

--- a/platforms/software/dependency-management/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadata.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadata.java
@@ -33,7 +33,10 @@ import java.io.File;
  * This should be better named as {@code OverriddenComponentIdArtifactMetadata} or similar, but Kotlin currently references this internal class.
  *
  * @see <a href="https://github.com/JetBrains/kotlin/blame/83a7ecda3eccd378a3882b81d8e83ecc12b3b786/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/targets/js/npm/resolver/KotlinCompilationNpmResolver.kt#L245">Link</a>
+ *
+ * @deprecated This class will be removed in Gradle 9.0.
  */
+@Deprecated
 public class CompositeProjectComponentArtifactMetadata implements LocalComponentArtifactMetadata, ComponentArtifactIdentifier {
 
     private final ComponentIdentifier overrideComponentId;

--- a/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentGraphResolveState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentGraphResolveState.java
@@ -18,7 +18,6 @@ package org.gradle.internal.component.local.model;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
-import org.gradle.api.Transformer;
 import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.artifacts.ModuleVersionIdentifier;
 import org.gradle.api.artifacts.component.ComponentIdentifier;
@@ -42,7 +41,6 @@ import org.gradle.internal.model.InMemoryCacheFactory;
 
 import javax.annotation.Nullable;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
@@ -60,9 +58,9 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
     private final ComponentIdGenerator idGenerator;
     private final boolean adHoc;
     private final LocalVariantGraphResolveStateFactory variantFactory;
-    private final Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> artifactTransformer;
     private final CalculatedValueContainerFactory calculatedValueContainerFactory;
     private final InMemoryCacheFactory cacheFactory;
+    private final ComponentIdentifier overrideComponentId;
 
     // The graph resolve state for variants selected by name
     private final InMemoryLoadingCache<String, LocalVariantGraphResolveState> variants;
@@ -82,7 +80,7 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
         LocalVariantGraphResolveStateFactory variantFactory,
         CalculatedValueContainerFactory calculatedValueContainerFactory,
         InMemoryCacheFactory cacheFactory,
-        @Nullable Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> artifactTransformer
+        @Nullable ComponentIdentifier overrideComponentId
     ) {
         super(instanceId, metadata, attributeDesugaring);
         this.idGenerator = idGenerator;
@@ -90,7 +88,7 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
         this.variantFactory = variantFactory;
         this.calculatedValueContainerFactory = calculatedValueContainerFactory;
         this.cacheFactory = cacheFactory;
-        this.artifactTransformer = artifactTransformer;
+        this.overrideComponentId = overrideComponentId;
 
         // Mutable state
         this.variants = cacheFactory.createCalculatedValueCache(Describables.of("variants"), this::doCreateLegacyConfiguration);
@@ -117,7 +115,7 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
         //       excludes to dependencies in a beforeResolve.
         this.graphSelectionCandidates.set(
             calculatedValueContainerFactory.create(Describables.of("variants of", getMetadata()), context ->
-                computeGraphSelectionCandidates(variantFactory, artifactTransformer)
+                computeGraphSelectionCandidates(variantFactory, overrideComponentId)
             )
         );
         this.selectableVariantResults.set(
@@ -138,16 +136,11 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
     }
 
     @Override
-    public LocalComponentGraphResolveState copy(ComponentIdentifier newComponentId, Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> transformer) {
-        // Keep track of transformed artifacts as a given artifact may appear in multiple variants
-        Map<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> transformedArtifacts = new HashMap<>();
-        Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> cachedTransformer = oldArtifact ->
-            transformedArtifacts.computeIfAbsent(oldArtifact, transformer::transform);
-
+    public LocalComponentGraphResolveState copyWithComponentId(ComponentIdentifier overrideComponentId) {
         LocalComponentGraphResolveMetadata originalMetadata = getMetadata();
         LocalComponentGraphResolveMetadata copiedMetadata = new LocalComponentGraphResolveMetadata(
             originalMetadata.getModuleVersionId(),
-            newComponentId,
+            overrideComponentId,
             originalMetadata.getStatus(),
             originalMetadata.getAttributesSchema()
         );
@@ -161,7 +154,7 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
             variantFactory,
             calculatedValueContainerFactory,
             cacheFactory,
-            cachedTransformer
+            overrideComponentId
         );
     }
 
@@ -179,14 +172,14 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
 
     private static LocalComponentGraphSelectionCandidates computeGraphSelectionCandidates(
         LocalVariantGraphResolveStateFactory variantFactory,
-        @Nullable Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> artifactTransformer
+        @Nullable ComponentIdentifier overrideComponentId
     ) {
         ImmutableList.Builder<LocalVariantGraphResolveState> variantsWithAttributes = new ImmutableList.Builder<>();
         ImmutableMap.Builder<String, LocalVariantGraphResolveState> variantsByConfigurationName = ImmutableMap.builder();
 
         variantFactory.visitConsumableVariants(variantState -> {
-            if (artifactTransformer != null) {
-                variantState = variantState.copyWithTransformedArtifacts(artifactTransformer);
+            if (overrideComponentId != null) {
+                variantState = variantState.copyWithComponentId(overrideComponentId);
             }
 
             if (!variantState.getAttributes().isEmpty()) {
@@ -238,8 +231,8 @@ public class DefaultLocalComponentGraphResolveState extends AbstractComponentGra
         if (variant == null) {
             return null;
         }
-        if (artifactTransformer != null) {
-            return variant.copyWithTransformedArtifacts(artifactTransformer);
+        if (overrideComponentId != null) {
+            return variant.copyWithComponentId(overrideComponentId);
         }
         return variant;
     }

--- a/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalVariantGraphResolveState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalVariantGraphResolveState.java
@@ -20,7 +20,6 @@ import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import org.gradle.api.artifacts.component.ComponentIdentifier;
 import org.gradle.api.internal.attributes.ImmutableAttributes;
-import org.gradle.composite.internal.CompositeProjectComponentArtifactMetadata;
 import org.gradle.internal.Describables;
 import org.gradle.internal.component.external.model.ImmutableCapabilities;
 import org.gradle.internal.component.model.ComponentArtifactMetadata;
@@ -80,10 +79,11 @@ public class DefaultLocalVariantGraphResolveState implements LocalVariantGraphRe
         ImmutableSet.Builder<LocalVariantMetadata> copiedArtifactSets = ImmutableSet.builderWithExpectedSize(artifactSets.size());
 
         for (LocalVariantMetadata oldArtifactSet : artifactSets) {
+            @SuppressWarnings("deprecation")
             CalculatedValue<ImmutableList<LocalComponentArtifactMetadata>> newArtifacts =
                 calculatedValueContainerFactory.create(Describables.of(oldArtifactSet.asDescribable(), "artifacts"), c ->
                     oldArtifactSet.getArtifacts().stream()
-                        .map(originalArtifact -> new CompositeProjectComponentArtifactMetadata(overrideComponentId, originalArtifact))
+                        .map(originalArtifact -> new org.gradle.composite.internal.CompositeProjectComponentArtifactMetadata(overrideComponentId, originalArtifact))
                         .collect(ImmutableList.toImmutableList())
                 );
 

--- a/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalVariantGraphResolveState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalVariantGraphResolveState.java
@@ -18,9 +18,9 @@ package org.gradle.internal.component.local.model;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
-import org.gradle.api.Transformer;
 import org.gradle.api.artifacts.component.ComponentIdentifier;
 import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.composite.internal.CompositeProjectComponentArtifactMetadata;
 import org.gradle.internal.Describables;
 import org.gradle.internal.component.external.model.ImmutableCapabilities;
 import org.gradle.internal.component.model.ComponentArtifactMetadata;
@@ -34,6 +34,7 @@ import org.gradle.internal.component.model.VariantResolveMetadata;
 import org.gradle.internal.model.CalculatedValue;
 import org.gradle.internal.model.CalculatedValueContainerFactory;
 
+import javax.annotation.Nullable;
 import java.util.List;
 import java.util.Set;
 
@@ -74,8 +75,7 @@ public class DefaultLocalVariantGraphResolveState implements LocalVariantGraphRe
     }
 
     @Override
-    public LocalVariantGraphResolveState copyWithTransformedArtifacts(Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> artifactTransformer) {
-
+    public LocalVariantGraphResolveState copyWithComponentId(ComponentIdentifier overrideComponentId) {
         Set<LocalVariantMetadata> artifactSets = artifactState.getArtifactVariants();
         ImmutableSet.Builder<LocalVariantMetadata> copiedArtifactSets = ImmutableSet.builderWithExpectedSize(artifactSets.size());
 
@@ -83,13 +83,13 @@ public class DefaultLocalVariantGraphResolveState implements LocalVariantGraphRe
             CalculatedValue<ImmutableList<LocalComponentArtifactMetadata>> newArtifacts =
                 calculatedValueContainerFactory.create(Describables.of(oldArtifactSet.asDescribable(), "artifacts"), c ->
                     oldArtifactSet.getArtifacts().stream()
-                        .map(artifactTransformer::transform)
+                        .map(originalArtifact -> new CompositeProjectComponentArtifactMetadata(overrideComponentId, originalArtifact))
                         .collect(ImmutableList.toImmutableList())
                 );
 
             copiedArtifactSets.add(new LocalVariantMetadata(
                 oldArtifactSet.getName(),
-                oldArtifactSet.getIdentifier(),
+                OverrideComponentIdArtifactSetIdentifier.of(overrideComponentId, oldArtifactSet.getIdentifier()),
                 oldArtifactSet.asDescribable(),
                 oldArtifactSet.getAttributes(),
                 oldArtifactSet.getCapabilities(),
@@ -218,5 +218,62 @@ public class DefaultLocalVariantGraphResolveState implements LocalVariantGraphRe
         public Set<LocalVariantMetadata> getArtifactVariants() {
             return artifactSets;
         }
+    }
+
+    /**
+     * Identifies an artifact set where the owning component identifier has been overridden.
+     * <p>
+     * When overriding the owning component identifier of an artifact, their identity changes. As such,
+     * the identity of the artifact set must also change. This identifier tracks the identity of an
+     * artifact set where the owning component identifier has been overridden, so that caches
+     * properly track the artifacts within this artifact set have an overridden component ID.
+     *
+     * @see org.gradle.internal.resolve.resolver.ResolvedVariantCache
+     */
+    private static class OverrideComponentIdArtifactSetIdentifier implements VariantResolveMetadata.Identifier {
+
+        private final ComponentIdentifier overrideComponentId;
+        private final VariantResolveMetadata.Identifier delegate;
+
+        public OverrideComponentIdArtifactSetIdentifier(
+            ComponentIdentifier overrideComponentId,
+            VariantResolveMetadata.Identifier delegate
+        ) {
+            this.overrideComponentId = overrideComponentId;
+            this.delegate = delegate;
+        }
+
+        @Nullable
+        public static OverrideComponentIdArtifactSetIdentifier of(
+            ComponentIdentifier overrideComponentId,
+            @Nullable VariantResolveMetadata.Identifier delegate
+        ) {
+            if (delegate == null) {
+                return null;
+            }
+
+            return new OverrideComponentIdArtifactSetIdentifier(overrideComponentId, delegate);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            OverrideComponentIdArtifactSetIdentifier that = (OverrideComponentIdArtifactSetIdentifier) o;
+            return overrideComponentId.equals(that.overrideComponentId) && delegate.equals(that.delegate);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = overrideComponentId.hashCode();
+            result = 31 * result + delegate.hashCode();
+            return result;
+        }
+
     }
 }

--- a/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalComponentGraphResolveState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalComponentGraphResolveState.java
@@ -16,7 +16,6 @@
 
 package org.gradle.internal.component.local.model;
 
-import org.gradle.api.Transformer;
 import org.gradle.api.artifacts.ModuleVersionIdentifier;
 import org.gradle.api.artifacts.component.ComponentIdentifier;
 import org.gradle.internal.component.model.ComponentGraphResolveState;
@@ -56,9 +55,9 @@ public interface LocalComponentGraphResolveState extends ComponentGraphResolveSt
     LocalComponentGraphResolveMetadata getMetadata();
 
     /**
-     * Copies this state, but with the new component ID and the artifacts transformed by the given transformer.
+     * Copies this state, but with the new component ID.
      */
-    LocalComponentGraphResolveState copy(ComponentIdentifier newComponentId, Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> transformer);
+    LocalComponentGraphResolveState copyWithComponentId(ComponentIdentifier newComponentId);
 
     /**
      * We currently allow a configuration that has been partially observed for resolution to be modified

--- a/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantGraphResolveState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantGraphResolveState.java
@@ -16,7 +16,7 @@
 
 package org.gradle.internal.component.local.model;
 
-import org.gradle.api.Transformer;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
 import org.gradle.internal.component.model.VariantGraphResolveState;
 
 import java.util.Set;
@@ -37,12 +37,12 @@ public interface LocalVariantGraphResolveState extends VariantGraphResolveState 
     Set<LocalFileDependencyMetadata> getFiles();
 
     /**
-     * Returns a copy of this variant, except with all artifacts transformed by the given transformer.
+     * Returns a copy of this variant, except with the given component identifier.
      *
-     * @param artifactTransformer A transformer applied to all artifacts and sub-variant artifacts.
+     * @param overrideComponentId The new owning component identifier.
      *
-     * @return A copy of this variant, with the given transformer applied to all artifacts.
+     * @return A copy of this variant, with the given owning component identifier.
      */
-    LocalVariantGraphResolveState copyWithTransformedArtifacts(Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> artifactTransformer);
+    LocalVariantGraphResolveState copyWithComponentId(ComponentIdentifier overrideComponentId);
 
 }

--- a/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantMetadata.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/LocalVariantMetadata.java
@@ -24,6 +24,8 @@ import org.gradle.internal.component.model.DefaultVariantMetadata;
 import org.gradle.internal.component.model.VariantResolveMetadata;
 import org.gradle.internal.model.CalculatedValue;
 
+import javax.annotation.Nullable;
+
 /**
  * Implementation of {@link VariantResolveMetadata} which allows variant artifacts to be calculated lazily
  * while holding a project lock.
@@ -31,7 +33,7 @@ import org.gradle.internal.model.CalculatedValue;
 public final class LocalVariantMetadata extends DefaultVariantMetadata {
     private final CalculatedValue<ImmutableList<LocalComponentArtifactMetadata>> artifacts;
 
-    public LocalVariantMetadata(String name, Identifier identifier, DisplayName displayName, ImmutableAttributes attributes, ImmutableCapabilities capabilities, CalculatedValue<ImmutableList<LocalComponentArtifactMetadata>> artifacts) {
+    public LocalVariantMetadata(String name, @Nullable Identifier identifier, DisplayName displayName, ImmutableAttributes attributes, ImmutableCapabilities capabilities, CalculatedValue<ImmutableList<LocalComponentArtifactMetadata>> artifacts) {
         super(name, identifier, displayName, attributes, ImmutableList.of(), capabilities);
         this.artifacts = artifacts;
     }

--- a/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeBuildsValueSnapshotterSerializerRegistry.java
+++ b/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeBuildsValueSnapshotterSerializerRegistry.java
@@ -21,6 +21,7 @@ import org.gradle.internal.snapshot.impl.ValueSnapshotterSerializerRegistry;
 
 public class CompositeBuildsValueSnapshotterSerializerRegistry extends DefaultSerializerRegistry implements ValueSnapshotterSerializerRegistry {
 
+    @SuppressWarnings("deprecation")
     public CompositeBuildsValueSnapshotterSerializerRegistry() {
         super();
         register(CompositeProjectComponentArtifactMetadata.class, new CompositeProjectComponentArtifactMetadataSerializer());

--- a/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadataSerializer.java
+++ b/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadataSerializer.java
@@ -24,6 +24,7 @@ import org.gradle.internal.serialize.Decoder;
 import org.gradle.internal.serialize.Encoder;
 import org.gradle.internal.serialize.Serializer;
 
+@SuppressWarnings("deprecation")
 public class CompositeProjectComponentArtifactMetadataSerializer implements Serializer<CompositeProjectComponentArtifactMetadata> {
 
     private final ComponentIdentifierSerializer componentIdentifierSerializer = new ComponentIdentifierSerializer();

--- a/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadataSerializer.java
+++ b/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeProjectComponentArtifactMetadataSerializer.java
@@ -24,8 +24,6 @@ import org.gradle.internal.serialize.Decoder;
 import org.gradle.internal.serialize.Encoder;
 import org.gradle.internal.serialize.Serializer;
 
-import java.io.File;
-
 public class CompositeProjectComponentArtifactMetadataSerializer implements Serializer<CompositeProjectComponentArtifactMetadata> {
 
     private final ComponentIdentifierSerializer componentIdentifierSerializer = new ComponentIdentifierSerializer();
@@ -35,14 +33,12 @@ public class CompositeProjectComponentArtifactMetadataSerializer implements Seri
     public CompositeProjectComponentArtifactMetadata read(Decoder decoder) throws Exception {
         ProjectComponentIdentifier componentIdentifier = (ProjectComponentIdentifier) componentIdentifierSerializer.read(decoder);
         PublishArtifactLocalArtifactMetadata delegate = publishArtifactLocalArtifactMetadataSerializer.read(decoder);
-        File file = new File(decoder.readString());
-        return new CompositeProjectComponentArtifactMetadata(componentIdentifier, delegate, file);
+        return new CompositeProjectComponentArtifactMetadata(componentIdentifier, delegate);
     }
 
     @Override
     public void write(Encoder encoder, CompositeProjectComponentArtifactMetadata value) throws Exception {
         componentIdentifierSerializer.write(encoder, value.getComponentIdentifier());
         publishArtifactLocalArtifactMetadataSerializer.write(encoder, (PublishArtifactLocalArtifactMetadata) value.getDelegate());
-        encoder.writeString(value.getFile().getCanonicalPath());
     }
 }

--- a/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultBuildTreeLocalComponentProvider.java
+++ b/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultBuildTreeLocalComponentProvider.java
@@ -29,12 +29,11 @@ import org.gradle.internal.Describables;
 import org.gradle.internal.build.CompositeBuildParticipantBuildState;
 import org.gradle.internal.build.IncludedBuildState;
 import org.gradle.internal.component.local.model.LocalComponentGraphResolveState;
-import org.gradle.internal.model.InMemoryLoadingCache;
 import org.gradle.internal.model.InMemoryCacheFactory;
+import org.gradle.internal.model.InMemoryLoadingCache;
 import org.gradle.util.Path;
 
 import javax.inject.Inject;
-import java.io.File;
 
 /**
  * Default implementation of {@link BuildTreeLocalComponentProvider}.
@@ -106,12 +105,7 @@ public class DefaultBuildTreeLocalComponentProvider implements BuildTreeLocalCom
         // Get the local component, then transform it to have a foreign identifier
         LocalComponentGraphResolveState originalComponent = originalComponents.get(projectIdentifier);
         ProjectComponentIdentifier foreignIdentifier = buildState.idToReferenceProjectFromAnotherBuild(projectIdentifier);
-        return originalComponent.copy(foreignIdentifier, originalArtifact -> {
-            // Currently need to resolve the file, so that the artifact can be used in both a script classpath and
-            // the main build. This accesses project state. Instead, the file should be resolved as required.
-            File file = projectState.fromMutableState(p -> originalArtifact.getFile());
-            return new CompositeProjectComponentArtifactMetadata(foreignIdentifier, originalArtifact, file);
-        });
+        return originalComponent.copyWithComponentId(foreignIdentifier);
     }
 
     @Override

